packages/flutter_tools/lib/src/compute_dev_dependencies.dart

```diff
@@ -2,126 +2,136 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-import 'base/logger.dart';
-import 'convert.dart';
-import 'dart/pub.dart';
+import 'package:file/file.dart';
+import 'package:package_config/package_config.dart';
+import 'package:yaml/yaml.dart';
+
 import 'project.dart';
 
-/// Returns dependencies of [project] that are _only_ used as `dev_dependency`.
+/// Computes a representation of the transitive dependency graph rooted at
+/// [project].
 ///
-/// That is, computes and returns a subset of dependencies, where the original
-/// set is based on packages listed as [`dev_dependency`][dev_deps] in the
-/// `pubspec.yaml` file, and removing packages from that set that appear as
-/// dependencies (implicitly non-dev) in any non-dev package depended on.
-Future<Set<String>> computeExclusiveDevDependencies(
-  Pub pub, {
-  required Logger logger,
-  required FlutterProject project,
-}) async {
-  final Map<String, Object?> jsonResult = await pub.deps(project);
-
-  Never fail([String? reason]) {
-    logger.printTrace(const JsonEncoder.withIndent('  ').convert(jsonResult));
-    throw StateError(
-      'dart pub deps --json ${reason != null ? 'had unexpected output: $reason' : 'failed'}',
-    );
-  }
-
-  List<T> asListOrFail<T>(Object? value, String name) {
-    // Allow omitting a list as empty to default to an empty list
-    if (value == null) {
-      return <T>[];
+/// Does a search rooted in `project`, following the `dependencies` and
+/// `dev_dependencies` of the pubspec to find the transitive dependencies of the
+/// app. (including the project itself). Will follow the `dev_dependencies` of
+/// the [project] package.
+///
+/// Will load each of the dependencies' `pubspec.yaml` using [fileSystem]. Using
+/// [packageConfig] to locate the files.
+///
+/// Avoids loading the [project] manifest again.
+///
+/// If a pubspec cannot be read, or is malformed, that pubspec is skipped. If
+/// nothing has changed since a succesful `pub get` that should never happen.
+/// In case of `flutter run --no-pub` this will at most ignore new dependencies.
+Map<String, Dependency> computeTransitiveDependencies(
+  FlutterProject project,
+  PackageConfig packageConfig,
+  FileSystem fileSystem, {
+  bool followDevDependencies = false,
+}) {
+  final Map<String, Dependency> result = <String, Dependency>{};
+
+  final List<String> packageNamesToVisit = <String>[project.manifest.appName];
+  while (packageNamesToVisit.isNotEmpty) {
```

Comment:
I think this loop can be refactored. It's quite large and has some very large conditional blocks that are hard to follow. Can the logic setting rootUri and pubspec be moved into a helper function that returns a record? Then the operations adding to the `dependencies` list can be handled separately.